hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/analysis/HoodieAnalysis.scala

```diff
@@ -128,15 +150,151 @@ object HoodieAnalysis {
     //       To work this around, we injecting this as the rule that trails pre-CBO, ie it's
     //          - Triggered before CBO, therefore have access to the same stats as CBO
     //          - Precedes actual [[customEarlyScanPushDownRules]] invocation
-    optimizerRules += (spark => HoodiePruneFileSourcePartitions(spark))
+    rules += (spark => HoodiePruneFileSourcePartitions(spark))
+
+    rules
+  }
+
+  /**
+   * This rule adjusts output of the [[LogicalRelation]] resolving int Hudi tables such that all of the
+   * default Spark resolution could be applied resolving standard Spark SQL commands
+   *
+   * <ul>
+   *  <li>`MERGE INTO ...`</li>
+   *  <li>`INSERT INTO ...`</li>
+   *  <li>`UPDATE ...`</li>
+   * </ul>
+   *
+   * even though Hudi tables might be carrying meta-fields that have to be ignored during resolution phase.
+   *
+   * Spark >= 3.2 bears fully-fledged support for meta-fields and such antics are not required for it:
+   * we just need to annotate corresponding attributes as "metadata" for Spark to be able to ignore it.
+   *
+   * In Spark < 3.2 however, this is worked around by simply removing any meta-fields from the output
+   * of the [[LogicalRelation]] resolving into Hudi table. Note that, it's a safe operation since we
+   * actually need to ignore these values anyway
+   */
+  case class AdaptIngestionTargetLogicalRelations(spark: SparkSession) extends Rule[LogicalPlan] {
```

Comment:
This is the core of the change:

 - Here we remove implementation of the bespoke resolution rules for Hudi components, instead relying on Spark to resolve these (most of these constructs don't have any custom logic relative to vanilla Spark SQL and therefore would be perfectly fine resolved by standard Spark resolution rules) 
 - Instead we repurpose these rules to serve as conversion point from Spark's standard constructs (like `MergeIntoTable`) into Hudi's ones that implement Hudi-specific semantic (`MergeIntoHoodieTableCommand`). Note that, we require these constructs be fully resolved prior to us trying to convert them